include/irr/builtin/glsl/bxdf/brdf/specular/geom/smith.glsl

```diff
@@ -35,86 +39,71 @@ float irr_glsl_GGXSmith_G1_wo_numerator(in float NdotX, in float a2, in float on
 {
-    float Vterm = NdotL * length(vec3(at*TdotV, ab*BdotV, NdotV));
-    float Lterm = NdotV * length(vec3(at*TdotL, ab*BdotL, NdotL));
+    float Vterm = NdotL*irr_glsl_smith_ggx_devsh_part(TdotV2,BdotV2,NdotV2,ax2,ay2);
+    float Lterm = NdotV*irr_glsl_smith_ggx_devsh_part(TdotL2,BdotL2,NdotL2,ax2,ay2);
     return 0.5 / (Vterm + Lterm);
 }
-
-float irr_glsl_ggx_smith_aniso_wo_numerator(in vec3 N, in vec3 L, in vec3 V, in vec3 T, in float NdotV, in float NdotL, in float ax2, in float ay2)
+float irr_glsl_ggx_smith_correlated(in float NdotV, in float TdotV2, in float BdotV2, in float NdotV2, in float NdotL, in float TdotL2, in float BdotL2, in float NdotL2, in float ax2, in float ay2)
 {
-    float a02 = irr_glsl_smith_aniso_a0_2(N, V, T, NdotV*NdotV, ax2, ay2);
-    float G1_V = irr_glsl_GGXSmith_G1_wo_numerator(NdotV, a02, 1.0-a02);
-    a02 = irr_glsl_smith_aniso_a0_2(N, L, T, NdotL*NdotL, ax2, ay2);
-    float G1_L = irr_glsl_GGXSmith_G1_wo_numerator(NdotL, a02, 1.0-a02);
-
-    return G1_V*G1_L;
-}
-float irr_glsl_ggx_smith_aniso(in vec3 N, in vec3 L, in vec3 V, in vec3 T, in float NdotV, in float NdotL, in float ax2, in float ay2)
-{
-    return 4.0*NdotV*NdotL*irr_glsl_ggx_smith_aniso_wo_numerator(N, L, V, T, NdotV, NdotL, ax2, ay2);
+    return 4.0*NdotV*NdotL*irr_glsl_ggx_smith_correlated_wo_numerator(NdotV, TdotV2, BdotV2, NdotV2, NdotL, TdotL2, BdotL2, NdotL2, ax2, ay2);
 }
 
-float irr_glsl_ggx_smith_correlated_wo_numerator(in vec3 N, in vec3 L, in vec3 V, in vec3 T, in float NdotV, in float NdotL, in float ax2, in float ay2)
+float irr_glsl_ggx_smith_G2_over_G1(in float NdotL, in float NdotL2, in float NdotV, in float NdotV2, in float a2, in float one_minus_a2)
 {
-    float a02 = irr_glsl_smith_aniso_a0_2(N, L, T, NdotL*NdotL, ax2, ay2);
-    float denom = NdotV*irr_glsl_smith_ggx_devsh_part(NdotL*NdotL,a02,1.0-a02);
-	a02 = irr_glsl_smith_aniso_a0_2(N, V, T, NdotV*NdotV, ax2, ay2);
-    denom += NdotL*irr_glsl_smith_ggx_devsh_part(NdotV*NdotV,a02,1.0-a02);
-    return 0.5 / denom;
+    float devsh_v = irr_glsl_smith_ggx_devsh_part(NdotV2,a2,one_minus_a2);
+	float G2_over_G1 = NdotL*(devsh_v + NdotV);
+	G2_over_G1 /= NdotV*irr_glsl_smith_ggx_devsh_part(NdotL2,a2,one_minus_a2) + NdotL*devsh_v;
+
+    return G2_over_G1;
 }
-float irr_glsl_ggx_smith_correlated(in vec3 N, in vec3 L, in vec3 V, in vec3 T, in float NdotV, in float NdotL, in float ax2, in float ay2)
+float irr_glsl_ggx_smith_G2_over_G1(in float NdotL, in float TdotL2, in float BdotL2, in float NdotL2, in float NdotV, in float TdotV2, in float BdotV2, in float NdotV2, in float ax2, in float ay2)
 {
-    return 4.0*NdotV*NdotL*irr_glsl_ggx_smith_correlated_wo_numerator(N, L, V, T, NdotV, NdotL, ax2, ay2);
-}
+    float devsh_v = irr_glsl_smith_ggx_devsh_part(TdotV2,BdotV2,NdotV2,ax2,ay2);
+	float G2_over_G1 = NdotL*(devsh_v + NdotV);
```

Comment:
I think there's supposed to be a factor of 2.0 here

P.S. we should make `irr_glsl_ggx_smith_G2_over_G1_wo_numerator`

Reply:
I think they'd be of the form `0.5*(1+devsh_v/NDotV)/(NdotL*devsh_v+NdotV*devsh_l)`

Reply:
mm or maybe a `irr_glsl_ggx_smith_G2_over_G1_wo_numerator_with_cos` could be more useful, cause it could have the `NdotL` factored in I suppose